### ISSUE TITLE
Add duk_opt_boolean(), duk_opt_number(), etc optional getters

### ISSUE THREAD

**Issue by svaarala** [svaarala/duktape#333]
Return required value or user-specified default if index or type is invalid. Example:

``` c
const char *message = duk_opt_string(ctx, 3 /*idx*/, "no message" /*default_value*/);
```

Equivalent to:

``` c
const char *message;
if (duk_is_string(ctx, 3)) {
    message = duk_get_string(ctx, 3);
} else {
    message = "no message";
}
```

**Comment by fatcerberus** [comment]
I like this idea, although I'm thinking the type checking should be stricter.  If a number is required and I pass a string, that should probably be a TypeError--but if the index is invalid or contains undefined, then I'd expect it to default.  Similar to how optional args are typically done in JS:

```javascript
if (typeof arg == "undefined") arg = defval;
// validate args
// do stuff
```

**Comment by svaarala** [comment]
There are actually several idioms and they're slightly different. For example:

```
if (x) {
    // matches if x is falsy: undefined, null, false, zero, NaN, empty string
}

if (x == null) {
    // matches if x is null or undefined (!)
}

if (x == undefined) {
    // matches if x is null or undefined
}

if (x === undefined) {
    // matches if x is undefined only
}

if (x === null) {
    // matches if x is null only
}

if (typeof x === 'undefined') {
    // matches undefined only, won't cause ReferenceError if 'x' is undeclared
}
```

Given these I'd perhaps favor the semantics that the default is used if:

- The index is invalid (which resembles an undefined variable case)
- The value is `null` or `undefined`, which are both usually handled together in Ecmascript semantics

So:

```c
if (!duk_is_valid_index(ctx, idx) || duk_is_undefined(ctx, idx) || duk_is_null(ctx, idx)) {
    /* default */
} else if (duk_is_string(ctx, idx)) {
    /* get value */
} else {
    /* TypeError */
}
```

**Comment by svaarala** [comment]
Anyway, what I mostly prefer when writing C bindings is a way to get e.g. a socket port value without type erroring if the argument has incorrect type.

Most Ecmascript code don't really do much of type checking, and you'll often see:

```js
function connect(host, port) {
    port = port || 80;
}
```

This will be obviously broken if port is, say, a buffer, but it's still common. For me it'd be OK if C code behaved the same way.

**Comment by fatcerberus** [comment]
Yeah, the thing is, when it comes to optional arguments (which I imagine this will mostly be used for) I generally find it unintuitive if the default value is silently used due to a non-matching type.  "Default if falsy" is probably too broad a check as well.

Another alternative that would be more intuitive than silently using the default in case of the wrong type would be to type-coerce the value to the correct type, and throw an error if that isn't possible.

I think it's okay in the long run if these are stricter than typical JS idioms, as they are only used by C code, which is expected to be trusted and should therefore have somewhat predictable behavior.

**Comment by svaarala** [comment]
I think I agree in the end, so the proposal so far would be for the semantics to match:

```c
if (!duk_is_valid_index(ctx, idx) || duk_is_undefined(ctx, idx) || duk_is_null(ctx, idx)) {
    /* default value */
} else if (duk_is_string(ctx, idx)) {
    /* get value */
} else {
    /* TypeError */
}
```

I somewhat dislike coercion behavior from a get API call. Even if the coercion wasn't done in-place (which would be very misleading) it would still potentially cause side effects one wouldn't expect from a plain get API call.

**Comment by svaarala** [comment]
Hmm, by the way, the reason I wouldn't expect `duk_opt_string()` to throw a TypeError is probably that `duk_get_string()` doesn't do that either. And "get string" and "get optional string" are pretty close in my mind so I expect it to silently default - just like `duk_get_string()` silently defaults to NULL for every invalid type.

**Comment by fatcerberus** [comment]
True, although to be fair I usually probably consider it an error condition if `duk_get_string()` returns NULL rather than treating it as a "default". :)

**Comment by svaarala** [comment]
So just to illustrate what I sometimes do:

```c
const char *msg;

msg = duk_get_string(ctx, idx);
if (!msg) {
    msg = "no message";
}
```

This never throws, and if the value at `idx` is a buffer, I get `"no message"`.

**Comment by svaarala** [comment]
The `NULL` return value just indicates that there's no appropriate value matching the type requested. So yes, it's not a "default" in that sense.

For callers that want a strict check there's `duk_require_string()` which is explicit in that there's a type check like the one discussed here.

So perhaps `duk_require_opt_string()` for throwing on invalid types and `duk_require_string()` for ignoring invalid types? :wink:

**Comment by svaarala** [comment]
Or another naming option: `duk_opt_xxx()` would indicate that the value is optional but that there's a type check.

Then add something like "get with default", which would be non-throwing like `duk_get_xxx()` but with an explicit (rather than implicit) default value. Say, `duk_get_string_default(ctx, idx, "no message")`. This was basically I was thinking about here: to make the default explicit.

**Comment by fatcerberus** [comment]
Hm, that would be ideal, I think.  I wasn't going to suggest two versions though because I thought it would have a detrimental effect on footprint.  But now that I think of it, in practice they could probably be implemented as macros, which would mitigate that.

**Comment by svaarala** [comment]
`duk_get_string(ctx, idx)` could be implemented through macros as `duk_get_string_default(ctx, idx, NULL)`. I tried that and it has a negligible impact on footprint. Of course, that's not ideal if most calls are without an explicit default.

There's an internal need for using explicit defaults in places though. Not sure if throwing or not throwing for invalid types would be the desired behavior there.

Actually what's needed internally here and there (and I presume user code) is convenience API calls for getting properties coerced/checked to a specific type, with a specified default. For example, for extending JX:

```c
duk_int_t max_depth = duk_get_prop_string_int(ctx, -3, "depth", 1000 /*default*/);
```

I've been considering adding these for a while - but there will be quite a lot of them. On the other hand if they're mostly macros it doesn't really matter all that much; it will still make calling code smaller.

**Comment by svaarala** [comment]
Another related API thing would be a multigetter with a format string, something like:

```c
/* Read .depth as integer, .loop as boolean */
duk_int_t depth = 1000;  /* assign defaults beforehand */
duk_bool_t loop = 0;
duk_get_prop_multi(ctx, -3, "depth:i loop:b", &depth, &loop);
```

There are similar questions here about how to default, when to throw, etc, and these API calls should be consistent and intuitive in what throws and when. That's why there's a clear separation between side effect free getters (`duk_get_xxx`) and side effecful coercers (`duk_to_xxx`) for example.

**Comment by fatcerberus** [comment]
Not really related, but you mentioning internal calls reminded me: What I really like about Duktape is that hacking on the engine itself isn't much different from using it outside: The implementation is so tight that the same API is used internally as well as externally, and so the codebase was already quite familiar to me by the time I actually needed to look at it. :)

**Comment by svaarala** [comment]
That's nice to hear :) Yeah, the hope there is that using the API internally too (not entirely but for a significant portion) would provide an incentive to improve the API also internally. Also performance measuring the engine also measures the API to some degree which is nice.

Looking for idioms / API calls to simplify internal code would be quite useful by the way. I've done some of that, but there's much room for improvement there. That multigetter for example would be quite useful internally in a couple of places. Similarly there'd be a couple of places where a "copy property list with attributes" primitive would be nice. There'd then be potential to expose them as an external API too, after "proving" they are generally useful and worth the footprint.

**Comment by fatcerberus** [comment]
I do the same thing with minisphere by the way, almost anything exposed to game scripts can also be used internally with identical functionality.  It really helps in simplifying the codebase and keeping it readable.

**Comment by svaarala** [comment]
Yeah, it's also nice if you can prototype a new call internally and if you're satisfied with it, only then expose it externally.

**Comment by svaarala** [comment]
There's actually two kinds of code inside Duktape in this respect. The first kind looks mostly like application code: it can look at internal structures, but mostly operates with the stack API, doesn't need to care about refcounts, etc. The second kind deals with `duk_tval`s, refcounts, and all the nitty gritty details that entails. It'd be nice if the second part was very small but it's not always possible :)

**Comment by svaarala** [comment]
There's now a duk_opt_xxx() in master (to be released in 2.1.0) which defaults for undefined/invalid index and TypeErrors for other non-matching types (including null), so closing.
